File: ros2_ws/src/my_cpp_pkg/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"

class NumberPublisherNode : public rclcpp::Node
{
public:
    NumberPublisherNode() : Node("number_publisher")
    {               
        this->declare_parameter("number", 2);
        this->declare_parameter("timer_period", 2.0);
        number_ = this->get_parameter("number").as_int();
        double timer_period = this->get_parameter("timer_period").as_double();

        param_callback_handle_ = this->add_post_set_parameters_callback(
            std::bind(&NumberPublisherNode::callbackParameters, this, std::placeholders::_1));
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        timer_ = this->create_wall_timer(std::chrono::duration<double>(timer_period), std::bind(&NumberPublisherNode::callbackPublish, this));        
        RCLCPP_INFO(this->get_logger(), "number_publisher node has been started");
    }
private:        
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    PostSetParametersCallbackHandle::SharedPtr param_callback_handle_;
    int number_;

    void callbackPublish()
    {
        auto msg = example_interfaces::msg::Int64();
        msg.data = number_;
        publisher_->publish(msg);
    }

    void callbackParameters(const std::vector<rclcpp::Parameter> & params)
    {
        for(const auto &param : params)
        {
            if(param.get_name() == "number") {
                number_ = param.as_int();
            }
        }
    }
};

int main(int argc, char **argv) 
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/src/my_py_pkg/my_py_pkg/number_publisher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from example_interfaces.msg import Int64

class NumberPublisherNode(Node):

    def __init__(self):
        super().__init__("number_publisher")
        self.declare_parameter("number", 2)
        self.declare_parameter("timer_period", 2.0)
        self.number_ = self.get_parameter("number").value
        self.publisher_ = self.create_publisher(Int64, "number", 10)
        self.timer_ = self.create_timer(self.get_parameter("timer_period").value, self.callback_publish)
        self.add_on_set_parameters_callback(self.callback_parameters)
        self.get_logger().info("number_publisher node has been started")

    def callback_publish(self):
        msg = Int64()
        msg.data = self.number_
        self.publisher_.publish(msg)

    def callback_parameters(self, params: list[Parameter]):
        for param in params:
            if param.name == "number":
                self.number_ = param.value


def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisherNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()